README.md
# Description
ZKMod is a Project Zomboid Steam Workshop Mod to collect and save player and world statistics on the server.

Workshop ID: 2733284288 / Mod ID: ZKMod

# Changelog
## ZKMod 1.1 (2022-02-02)
* Feature: world(_history).csv: added extended weather data
* Feature: player_(name).csv: added isFemale
* Feature: events_history.csv: Saving player perk level ups (multiline)
* Feature: players_online.csv: Players currently online (single line) 

## ZKMod 1.0.1 (2022-01-28)
* Fix: worldData.worldAgeDays incorrect
* new preview image

## ZKMod 1.0 (2022-01-27)
* collects player and game world statistics
* saves collected data to CSV files on the server

Changelog.md
ZKMod 1.0 (2022-27-01)
- collects player and game world statistics
- saves collected data to CSV files on the server

media/lua/client/ZKModClient.lua
-- Made by IkeC
-- CSV saving based on "Server Players Data" by Lemos: https://steamcommunity.com/sharedfiles/filedetails/?id=2695733462

require "ZKModShared"

if isServer() then
    return
end


local function ZKGetCommonPlayerData()
    -- https://zomboid-javadoc.com/41.65/zombie/characters/IsoPlayer.html
    local player = getPlayer()

    if not player then
        print("ZKModClient: ZKGetCommonPlayerData: No player data, exiting")
        return
    end

    -- https://zomboid-javadoc.com/41.65/zombie/GameTime.html
    local gametime = getGameTime()

    local username = player:getUsername()
    local forname = player:getDescriptor():getForename()
    local surname = player:getDescriptor():getSurname()

    local playerData = {}

    playerData.systemDate = ZKGetSystemDate()
    playerData.systemTime = ZKGetSystemTime()
    playerData.gametime = ZKGetGameTimeString(gametime)
    playerData.steamID = getSteamIDFromUsername(username)

    playerData.username = username
    playerData.charName = forname .. " " .. surname
    return playerData
end

-- Called on the player to parse its player data and send it to the server every ten (in-game) minutes
local function SendPlayerData(isdead)

    print("ZKModClient: SendPlayerData: isdead=" .. tostring(isdead))

    -- https://zomboid-javadoc.com/41.65/zombie/characters/IsoPlayer.html
    local player = getPlayer()
    if not player then
        print("ZKModClient: SendPlayerData: No player data, exiting")
        return
    end

    local playerData = ZKGetCommonPlayerData()

    if playerData then
        playerData.x = string.format("%.1f", player:getX())
        playerData.y = string.format("%.1f", player:getY())
        playerData.z = string.format("%.1f", player:getZ())

        playerData.isAlive = player:isAlive()
        playerData.zombieKills = player:getZombieKills()
        playerData.hoursSurvived = string.format("%.1f", player:getHoursSurvived())
        playerData.timeSurvived = player:getTimeSurvived()

        -- https://zomboid-javadoc.com/41.65/zombie/characters/BodyDamage/Nutrition.html
        local nutrition = player:getNutrition()
        playerData.weight = string.format("%.1f", nutrition:getWeight())
        playerData.characterHaveWeightTrouble = nutrition:characterHaveWeightTrouble()

        -- https://zomboid-javadoc.com/41.65/zombie/characters/SurvivorDesc.html
        local descriptor = player:getDescriptor()
        playerData.profession = descriptor:getProfession()

        -- https://zomboid-javadoc.com/41.65/zombie/characters/traits/TraitCollection.html
        local traits = player:getTraits()
        playerData.traits = traits:toString()

        playerData.isFemale = player:isFemale()

        local command = "SendPlayerDataAlive"
        if isdead then
            command = "SendPlayerDataDead"
        end

        print("ZKModClient: " .. playerData.username .. ": command=" .. command)

        sendClientCommand(player, "ZKMod", command, playerData)
    end
end

local function SendPlayerDataAlive()
    SendPlayerData(false)
end

local function SendPlayerDataDead()
    SendPlayerData(true)
end

Events.EveryHours.Add(SendPlayerDataAlive)
Events.OnPlayerDeath.Add(SendPlayerDataDead)


-- https://pzwiki.net/wiki/Modding:Lua_Events/LevelPerk
-- IsoGameCharacter The character whose perk is being leveled up or down.
-- https://zomboid-javadoc.com/41.65/zombie/characters/skills/PerkFactory.Perk.html The perk being leveled up or down.
-- Integer Perk level.
-- Boolean Whether the perk is being leveled up.
local function ZKLevelPerk(character, perk, level, levelUp)
    print("ZKModClient.ZKLevelPerk")

    -- https://zomboid-javadoc.com/41.65/zombie/characters/IsoPlayer.html
    local player = getPlayer()
    if not player then
        print("ZKModClient: SendPlayerData: No player data, exiting")
        return
    end
    
    local eventData = ZKGetCommonPlayerData()
    if eventData then
        eventData.eventName = "LevelPerk"
        eventData.eventData1 = perk:getName()
        eventData.eventData2 = level
        eventData.eventData3 = levelUp
        eventData.eventData4 = ""
        eventData.eventData5 = ""

        sendClientCommand(player, "ZKMod", "LevelPerk", eventData)
    end
end
Events.LevelPerk.Add(ZKLevelPerk)

--https://pzwiki.net/wiki/Modding:Lua_Events/OnNewFire
--https://pzwiki.net/wiki/Modding:Lua_Events/OnZombieDead
--https://pzwiki.net/wiki/Modding:Lua_Events/OnCharacterMeet

media/lua/shared/ZKModShared.lua
-- Made by IkeC
-- CSV saving based on "Server Players Data" by Lemos: https://steamcommunity.com/sharedfiles/filedetails/?id=2695733462

function ZKGetGameTimeString(gt)
    local minutes = string.format("%02d", gt:getMinutes())
    local result = gt:getDay()+1 .. "." .. gt:getMonth()+1 .. "." .. gt:getYear() .. 
        " " .. gt:getHour() .. ":" .. minutes        

    return result
end

function ZKGetCSVHeader(data)
    local strHeader = ""
    local separator = ""
    
    -- build the header
    for k,v in pairs(data) do
        if strHeader ~= "" then
            separator = ";"
        end

        strHeader = strHeader .. separator .. k
    end
    strHeader = strHeader .. "\n"

    return strHeader
end

function ZKGetCSVLine(data)
    local strData = ""
    local separator = ""

    -- fill data on next line
    for k,v in pairs(data) do
        if strData ~= "" then
            separator = ";"
        end
        
        if type(v)=="string" then
            strData = strData .. separator ..  "\"" .. v .. "\""
        else
            strData = strData .. separator .. tostring(v)
        end
    end
   
    strData = strData .. "\n"

    return strData
end

function ZKDump(o)
    if type(o) == 'table' then
       local s = '{ '
       for k,v in pairs(o) do
          if type(k) ~= 'number' then k = '"'..k..'"' end
          s = s .. '['..k..'] = ' .. ZKDump(v) .. ','
       end
       return s .. '} '
    else
       return tostring(o)
    end
 end

 function ZKGetSystemDate()
    return tostring(os.date("%d.%m.%Y",  os.time() + 1 * 60 * 60))
 end

 function ZKGetSystemTime()
    return tostring(os.date("%H:%M:%S",  os.time() + 1 * 60 * 60))
 end

media/lua/server/ZKModServer.lua
-- Made by IkeC
-- CSV saving based on "Server Players Data" by Lemos: https://steamcommunity.com/sharedfiles/filedetails/?id=2695733462

require "ZKModShared"

print("ZKModServer: isServer=" .. tostring(isServer()))

if not isServer() then
    return
end

-- Parse player data and save it to a .csv file inside Lua/ZKMod/ folder
local function SavePlayerData(data, savedead)
    if data then

        print("ZKModServer: data.isAlive=" .. tostring(data.isAlive) .. " savedead=" .. tostring(savedead))

        -- overwrite client data with server date/time
        data.systemDate = ZKGetSystemDate()
        data.systemTime = ZKGetSystemTime()

        local strHeader = ZKGetCSVHeader(data)
        local strData = ZKGetCSVLine(data)

        local filePath = "/ZKMod/player_" .. data.username .. ".csv"
        local dataFile = getFileWriter(filePath, true, false)
        dataFile:write(strHeader)
        dataFile:write(strData)
        dataFile:close()

        print("ZKModServer: SavePlayerData: " .. data.username .. " data saved to file")

        if savedead and (not data.isAlive) then
            print("ZKModServer: SavePlayerData: " .. data.username .. " dead, saving death")
            filePath = "/ZKMod/deaths.csv"
            dataFile = getFileWriter(filePath, true, true)
            dataFile:write(strData)
            dataFile:close()
        end
    end
end

-- Parse player event data and save it to event_history.csv file inside Lua/ZKMod/ folder
local function SaveEventData(data)
    if data then
        -- overwrite client data with server date/time
        data.systemDate = ZKGetSystemDate()
        data.systemTime = ZKGetSystemTime()

        print("ZKModServer: SaveEventData=" .. ZKDump(data))
        local strData = ZKGetCSVLine(data)

        local filePath = "/ZKMod/event_history.csv"
        local dataFile = getFileWriter(filePath, true, true)

        dataFile:write(strData)
        dataFile:close()
    end
end

-- executed when a client(player) sends its information to the server
local PlayerDataReceived = function(module, command, player, args)
    if module ~= "ZKMod" then
        return;
    end

    if command == "SendPlayerDataAlive" then
        print("ZKModServer: " .. command .. " for " .. args.username .. " received")
        SavePlayerData(args, false)
    end

    if command == "SendPlayerDataDead" then
        print("ZKModServer: " .. command .. " for " .. args.username .. " received")
        SavePlayerData(args, true)
    end

    if command == "LevelPerk" then
        print("ZKModServer: " .. command .. " for " .. args.username .. " received")
        SaveEventData(args)
    end
end

-- executed on server every in-game-hour
local ZKEveryHour = function(module, command, player, args)
    -- print("ZKModServer.ZKEveryHour")

    local worldData = ZKGetWorldData()

    -- print("ZKModServer.ZKEveryHour: worldData: " .. ZKDump(worldData))

    local strData = ZKGetCSVLine(worldData)
    print("ZKModServer.ZKEveryHour: strData: " .. strData)

    local filePath = "/ZKMod/world_history.csv"

    local dataFile = getFileWriter(filePath, true, true)
    dataFile:write(strData)
    dataFile:close()

    print("ZKModServer.ZKEveryHour: saved to " .. filePath)
end

-- executed on server every 10 in-game-minutes
local ZKEveryTenMinutes = function(module, command, player, args)
    -- print("ZKModServer.ZKEveryTenMinutes")

    local worldData = ZKGetWorldData()

    -- print("ZKModServer.ZKEveryTenMinutes: worldData: " .. ZKDump(worldData))

    local strHeader = ZKGetCSVHeader(worldData)
    local strData = ZKGetCSVLine(worldData)

    local filePath = "/ZKMod/world.csv"
    local dataFile = getFileWriter(filePath, true, false)

    dataFile:write(strHeader)
    dataFile:write(strData)
    dataFile:close()

    print("ZKModServer.ZKEveryTenMinutes: saved to " .. filePath)

    
    local playerData = ZKGetOnlinePlayers()

    -- print("ZKModServer.ZKEveryHour: worldData: " .. ZKDump(worldData))

    filePath = "/ZKMod/players_online.csv"

    local dataFile = getFileWriter(filePath, true, false)
    dataFile:write(playerData)
    dataFile:close()

    print("ZKModServer.ZKEveryTenMinutes: saved to " .. filePath)
end

-- get game world related data
function ZKGetWorldData()
    local worldData = {}

    -- https://zomboid-javadoc.com/41.65/zombie/GameTime.html
    local gt = getGameTime()
    local world = getWorld()

    worldData.serverDate = ZKGetSystemDate()
    worldData.serverTime = ZKGetSystemTime()
    worldData.gameTime = ZKGetGameTimeString(gt)
    worldData.worldAgeDays = math.floor(world:getWorldAgeDays())
    worldData.globalTemperature = string.format("%.1f", world:getGlobalTemperature())

    local onlinePlayers = getOnlinePlayers()
    local onlinePlayersCount = 0
    if onlinePlayers then
        onlinePlayersCount = onlinePlayers:size()
    end
    worldData.onlinePlayersCount = onlinePlayersCount

    worldData.isGamePaused = isGamePaused()
    worldData.night = string.format("%.1f", gt:getNight())

    local weatherData = ZKGetWeather()
    -- joining data: https://stackoverflow.com/a/1283399 
    for k,v in pairs(weatherData) do worldData[k] = v end

    return worldData
end

function ZKGetOnlinePlayers()
    local result = ""    
    local oPlayers = getOnlinePlayers()
    if oPlayers then
        local players = oPlayers:clone()
        for i = 0, players:size() - 1 do
            result = result .. players:get(i):getUsername()
            if i < players:size() - 1 then
                result = result .. ";"
            end
        end
    end
    return result
end

local function ZKWriteEvent(username, message)
    print("ZKModServer.ZKWriteEvent: username=" .. username .. " message=" .. message)
    local data = {}
    local gt = getGameTime()

    data.serverDate = ZKGetSystemDate()
    data.serverTime = ZKGetSystemTime()
    data.gameTime = ZKGetGameTimeString(gt)
    data.username = username
    data.message = message

    local strData = ZKGetCSVLine(data)
    local filePath = "/ZKMod/events.csv"

    local dataFile = getFileWriter(filePath, true, true)
    dataFile:write(strData)
    dataFile:close()
end

-- https://pzwiki.net/wiki/Modding:Lua_Events
Events.OnClientCommand.Add(PlayerDataReceived)

Events.EveryHours.Add(ZKEveryHour)
Events.EveryTenMinutes.Add(ZKEveryTenMinutes)

-- https://pzwiki.net/wiki/Modding:Lua_Events/OnRainStart
function ZKOnRainStart()
    print("ZKModServer.ZKOnRainStart")
    -- ZKWriteEvent("global", "Es beginnt zu regnen.")
end
Events.OnRainStart.Add(ZKOnRainStart)

-- https://pzwiki.net/wiki/Modding:Lua_Events/OnRainStop
function ZKOnRainStop()
    print("ZKModServer.ZKOnRainStop")
    -- ZKWriteEvent("global", "Der Regen hat aufgehört.")
end
Events.OnRainStop.Add(ZKOnRainStop)

-- https://pzwiki.net/wiki/Modding:Lua_Events/OnDawn
function ZKOnDawn()
    print("ZKModServer.ZKOnDawn")
    -- ZKWriteEvent("global", "Die Sonne geht auf.")
end
Events.OnDawn.Add(ZKOnDawn)

-- https://pzwiki.net/wiki/Modding:Lua_Events/OnDusk
function ZKOnDusk()
    print("ZKModServer.ZKOnDusk")
    -- ZKWriteEvent("global", "Die Sonne geht unter.")
end
Events.OnDusk.Add(ZKOnDusk)

-- https://pzwiki.net/wiki/Modding:Lua_Events/OnChangeWeather
-- String A string representing the weather. Can be either: "normal", "cloud", "rain", or "sunny"
function ZKOnChangeWeather(weather)
    print("ZKModServer.ZKOnChangeWeather")
    print("ZKModServer.ZKOnChangeWeather: weather=" .. weather)
end
Events.OnChangeWeather.Add(ZKOnChangeWeather)

function ZKOnWeatherPeriodStart(weatherPeriod)
    print("ZKModServer.ZKOnWeatherPeriodStart")
    --print(weatherPeriod)
end
Events.OnWeatherPeriodStart.Add(ZKOnWeatherPeriodStart)

-- works! (called twice?)
function ZKOnWeatherPeriodStop(weatherPeriod)
    print("ZKModServer.ZKOnWeatherPeriodStop")
    --print(weatherPeriod)
end
Events.OnWeatherPeriodStop.Add(ZKOnWeatherPeriodStop)

function ZKOnWeatherPeriodComplete(weatherPeriod)
    print("ZKModServer.ZKOnWeatherPeriodComplete")
    --print(weatherPeriod)
end
Events.OnWeatherPeriodComplete.Add(ZKOnWeatherPeriodComplete)

-- works!
function ZKOnWeatherPeriodStage(weatherPeriod)
    -- https://zomboid-javadoc.com/41.65/zombie/iso/weather/WeatherPeriod.html
    print("ZKModServer.ZKOnWeatherPeriodStage")
    print(weatherPeriod)
end
Events.OnWeatherPeriodStage.Add(ZKOnWeatherPeriodStage)

Events.OnRainStart.Add(ZKOnRainStart)
Events.OnRainStop.Add(ZKOnRainStop)


-- Based on weather function by Snake: http://pzmodding.blogspot.com/
function ZKGetWeather()
    local gt = GameTime:getInstance()
    local c = getClimateManager()

	local clouds = c:getCloudIntensity()
	local fog = c:getFogIntensity()
	local windpower = c:getWindPower()
	local windspeed = c:getWindspeedKph()
	local precipitationIntensity = c:getPrecipitationIntensity()

    local weatherData = {}
    weatherData.cloudIntensity = 0
    weatherData.fogIntensity = 0
    weatherData.precipitationIntensity = 0

    weatherData.windpower = 0
    weatherData.windspeed = 0

    weatherData.rain = false
    weatherData.snow = false

    weatherData.thunderstorm = false
    weatherData.tropicalstorm = false
    weatherData.blizzard = false
    
    -- wind
    if windpower > 0 then 
        weatherData.windpower = string.format("%.1f", windpower)
    end 
    if windspeed > 0 then
        weatherData.windspeed = string.format("%.1f", windspeed)        
    end

    if RainManager.isRaining() or (c:getPrecipitationIntensity() and c:getPrecipitationIntensity() > 0) then
        -- precipitation (snow, rain, storm...)
        weatherData.precipitationIntensity = string.format("%.1f", precipitationIntensity)

		if c:getPrecipitationIsSnow() then
            weatherData.snow = true
        else 
			if c:getWeatherPeriod():isRunning() then
				local wp = c:getWeatherPeriod()
				if wp:isThunderStorm() then
                    weatherData.thunderstorm = true
				elseif wp:isTropicalStorm() then
                    weatherData.tropicalstorm = true
				elseif wp:isBlizzard() then
                    weatherData.blizzard = true
				end
			end
            if weatherData.thunderstorm == false and weatherData.tropicalstorm == false and weatherData.blizzard == false then
             -- plain rain
             weatherData.rain = true
            end 
		end
	elseif (not RainManager.isRaining()) then
		-- cloudy without rain
        if clouds > 0 then
            weatherData.cloudIntensity = string.format("%.1f", clouds)
        end
        if fog > 0 then
            weatherData.fogIntensity = string.format("%.1f", fog)
        end

        if c:getPrecipitationIntensity() and c:getPrecipitationIntensity() > 0 and c:getPrecipitationIsSnow() then 
            --snow
            weatherData.snow = true
        end
	end

    return weatherData
end
